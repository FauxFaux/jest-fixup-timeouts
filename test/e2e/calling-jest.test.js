const fs = require('fs');
const process = require('child_process');
const tmp = require('tmp');
const strip = require('strip-ansi');

describe('the results of calling jest', () => {
  let dir, removeCallback;

  it('does nothing by default', async () => {
    expect(
      jestMessages(
        // language=JavaScript
        'const sleep = require("sleep-promise");\n' +
          'test("foo", async () => {\n' +
          ' await sleep(60);\n' +
          '}, 30);',
        {},
      ),
    ).toMatchSnapshot();
  });

  it('detects deadline exceeded for a test', async () => {
    expect(
      jestMessages(
        // language=JavaScript
        'const sleep = require("sleep-promise");\n' +
          'test("foo", async () => {\n' +
          ' await sleep(60);\n' +
          '}, 30);',
        {
          testEnvironment: '../../environment',
        },
        {
          plugins: ['../../rewrite-awaits'],
        },
      ),
    ).toMatchSnapshot();
  });

  function jestMessages(testJavaScript, jestConfig, babelConfig = {}) {
    fs.writeFileSync(`${dir}/foo.test.js`, testJavaScript);
    fs.writeFileSync(
      `${dir}/jest.config.json`,
      JSON.stringify({
        testFailureExitCode: 0,
        testRunner: 'jest-circus/runner',
        ...jestConfig,
      }),
    );
    fs.writeFileSync(
      `${dir}/babel.config.json`,
      JSON.stringify({
        ...babelConfig,
      }),
    );
    const stdout = runJest(
      dir,
      ['--runInBand', '--json', '--no-colors', '--no-cache'],
      jestConfig,
    );
    return JSON.parse(stdout.toString('utf-8')).testResults.map((result) =>
      strip(result.message),
    );
  }

  beforeEach(() => {
    const cache = `${__dirname}/../../.temp`;
    if (!fs.existsSync(cache)) {
      fs.mkdirSync(cache);
    }
    ({ name: dir, removeCallback } = tmp.dirSync({
      tmpdir: cache,
      unsafeCleanup: true,
    }));
  });

  afterEach(() => {
    removeCallback();
  });
});

function runJest(dir, args) {
  try {
    return process.execFileSync(require.resolve('jest/bin/jest'), args, {
      stdio: ['ignore', 'pipe', 'pipe'],
      cwd: dir,
    });
  } catch (err) {
    console.info(err.stdout.toString('utf-8'));
    console.error(err.stderr.toString('utf-8'));
    throw err;
  }
}
