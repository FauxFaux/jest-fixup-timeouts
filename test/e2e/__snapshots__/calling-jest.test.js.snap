// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the results of calling jest detects deadline exceeded for a test 1`] = `
Array [
  "  ● foo

    deadline exceeded (waited here for 9ms)

      1 | const sleep = require(\\"sleep-promise\\");
      2 | test(\\"foo\\", async () => {
    > 3 |  await sleep(60);
        |  ^
      4 | }, 30);

      at Object.<anonymous> (foo.test.js:3:2)
",
]
`;

exports[`the results of calling jest does nothing by default 1`] = `
Array [
  "  ● foo

    thrown: \\"Exceeded timeout of 30 ms for a test.
    Use jest.setTimeout(newTimeout) to increase the timeout value, if this is a long-running test.\\"

      1 | const sleep = require(\\"sleep-promise\\");
    > 2 | test(\\"foo\\", async () => {
        | ^
      3 |  await sleep(60);
      4 | }, 30);

      at Object.<anonymous> (foo.test.js:2:1)
",
]
`;
